### PR TITLE
ByteBuffer Implementation of FileWordList Cache

### PR DESCRIPTION
**Pull request by serac**
The primary benefit of these changes is that the memory usage in bytes is a clear function of the _cachePercent_ parameter. In the best case it is what one would expect from a parameter named _cachePercent_, `fileSize * 100 / cachePercent` and in the worst case (from my testing) `2 * fileSize * 100 / cachePercent`. That's at least 50% less than the current TreeMap implementation given that you're storing a 4-byte key along with the 8-byte file offset for every cached value.

The effect on performance is slightly beneficial. The following results are for 10 runs of the new implementation at the default cache size versus 10 runs of the current implementation at default cache size:

ByteBuffer Implementation@5%

    Test Run #1
    WordListDictionary (FileWordList) time to construct: 74ms
    WordListDictionary (FileWordList) total search time: 1ms
    WordListDictionary (FileWordList) avg time per search: 111ns
    WordListDictionary (MemoryMappedFileWordList) time to construct: 80ms
    WordListDictionary (MemoryMappedFileWordList) total search time: 0ms
    WordListDictionary (MemoryMappedFileWordList) avg time per search: 57ns

    Test Run #2
    WordListDictionary (MemoryMappedFileWordList) time to construct: 86ms
    WordListDictionary (FileWordList) time to construct: 86ms
    WordListDictionary (FileWordList) total search time: 1ms
    WordListDictionary (FileWordList) avg time per search: 134ns
    WordListDictionary (MemoryMappedFileWordList) total search time: 0ms
    WordListDictionary (MemoryMappedFileWordList) avg time per search: 72ns

    Test Run #3
    WordListDictionary (FileWordList) time to construct: 76ms
    WordListDictionary (FileWordList) total search time: 1ms
    WordListDictionary (FileWordList) avg time per search: 161ns
    WordListDictionary (MemoryMappedFileWordList) time to construct: 86ms
    WordListDictionary (MemoryMappedFileWordList) total search time: 1ms
    WordListDictionary (MemoryMappedFileWordList) avg time per search: 163ns

    Test Run #4
    WordListDictionary (FileWordList) time to construct: 85ms
    WordListDictionary (FileWordList) total search time: 1ms
    WordListDictionary (FileWordList) avg time per search: 105ns
    WordListDictionary (MemoryMappedFileWordList) time to construct: 89ms
    WordListDictionary (MemoryMappedFileWordList) total search time: 0ms
    WordListDictionary (MemoryMappedFileWordList) avg time per search: 54ns

    Test Run #5
    WordListDictionary (FileWordList) time to construct: 80ms
    WordListDictionary (MemoryMappedFileWordList) time to construct: 82ms
    WordListDictionary (FileWordList) total search time: 1ms
    WordListDictionary (FileWordList) avg time per search: 118ns
    WordListDictionary (MemoryMappedFileWordList) total search time: 0ms
    WordListDictionary (MemoryMappedFileWordList) avg time per search: 58ns

    Test Run #6
    WordListDictionary (MemoryMappedFileWordList) time to construct: 84ms
    WordListDictionary (MemoryMappedFileWordList) total search time: 0ms
    WordListDictionary (MemoryMappedFileWordList) avg time per search: 58ns
    WordListDictionary (FileWordList) time to construct: 86ms
    WordListDictionary (FileWordList) total search time: 1ms
    WordListDictionary (FileWordList) avg time per search: 113ns

    Test Run #7
    WordListDictionary (FileWordList) time to construct: 85ms
    WordListDictionary (FileWordList) total search time: 0ms
    WordListDictionary (FileWordList) avg time per search: 95ns
    WordListDictionary (MemoryMappedFileWordList) time to construct: 93ms
    WordListDictionary (MemoryMappedFileWordList) total search time: 0ms
    WordListDictionary (MemoryMappedFileWordList) avg time per search: 52ns

    Test Run #8
    WordListDictionary (FileWordList) time to construct: 80ms
    WordListDictionary (MemoryMappedFileWordList) time to construct: 85ms
    WordListDictionary (MemoryMappedFileWordList) total search time: 0ms
    WordListDictionary (MemoryMappedFileWordList) avg time per search: 70ns
    WordListDictionary (FileWordList) total search time: 1ms
    WordListDictionary (FileWordList) avg time per search: 113ns

    Test Run #9
    WordListDictionary (FileWordList) time to construct: 83ms
    WordListDictionary (FileWordList) total search time: 1ms
    WordListDictionary (FileWordList) avg time per search: 104ns
    WordListDictionary (MemoryMappedFileWordList) time to construct: 93ms
    WordListDictionary (MemoryMappedFileWordList) total search time: 0ms
    WordListDictionary (MemoryMappedFileWordList) avg time per search: 53ns

    Test Run #10
    WordListDictionary (FileWordList) time to construct: 81ms
    WordListDictionary (MemoryMappedFileWordList) time to construct: 85ms
    WordListDictionary (FileWordList) total search time: 1ms
    WordListDictionary (FileWordList) avg time per search: 107ns
    WordListDictionary (MemoryMappedFileWordList) total search time: 0ms
    WordListDictionary (MemoryMappedFileWordList) avg time per search: 52ns

TreeMap Implementation@5%

    Test Run #1
    WordListDictionary (FileWordList) time to construct: 93ms
    WordListDictionary (MemoryMappedFileWordList) time to construct: 95ms
    WordListDictionary (MemoryMappedFileWordList) total search time: 0ms
    WordListDictionary (MemoryMappedFileWordList) avg time per search: 71ns
    WordListDictionary (FileWordList) total search time: 1ms
    WordListDictionary (FileWordList) avg time per search: 148ns

    Test Run #2
    WordListDictionary (FileWordList) time to construct: 82ms
    WordListDictionary (FileWordList) total search time: 1ms
    WordListDictionary (FileWordList) avg time per search: 112ns
    WordListDictionary (MemoryMappedFileWordList) time to construct: 87ms
    WordListDictionary (MemoryMappedFileWordList) total search time: 0ms
    WordListDictionary (MemoryMappedFileWordList) avg time per search: 62ns

    Test Run #3
    WordListDictionary (FileWordList) time to construct: 89ms
    WordListDictionary (MemoryMappedFileWordList) time to construct: 96ms
    WordListDictionary (MemoryMappedFileWordList) total search time: 0ms
    WordListDictionary (MemoryMappedFileWordList) avg time per search: 82ns
    WordListDictionary (FileWordList) total search time: 1ms
    WordListDictionary (FileWordList) avg time per search: 137ns

    Test Run #4
    WordListDictionary (FileWordList) time to construct: 91ms
    WordListDictionary (FileWordList) total search time: 1ms
    WordListDictionary (FileWordList) avg time per search: 125ns
    WordListDictionary (MemoryMappedFileWordList) time to construct: 101ms
    WordListDictionary (MemoryMappedFileWordList) total search time: 0ms
    WordListDictionary (MemoryMappedFileWordList) avg time per search: 61ns

    Test Run #5
    WordListDictionary (FileWordList) time to construct: 75ms
    WordListDictionary (FileWordList) total search time: 1ms
    WordListDictionary (FileWordList) avg time per search: 115ns
    WordListDictionary (MemoryMappedFileWordList) time to construct: 83ms
    WordListDictionary (MemoryMappedFileWordList) total search time: 0ms
    WordListDictionary (MemoryMappedFileWordList) avg time per search: 62ns

    Test Run #6
    WordListDictionary (MemoryMappedFileWordList) time to construct: 105ms
    WordListDictionary (MemoryMappedFileWordList) total search time: 0ms
    WordListDictionary (MemoryMappedFileWordList) avg time per search: 68ns
    WordListDictionary (FileWordList) time to construct: 121ms
    WordListDictionary (FileWordList) total search time: 1ms
    WordListDictionary (FileWordList) avg time per search: 152ns

    Test Run #7
    WordListDictionary (FileWordList) time to construct: 91ms
    WordListDictionary (FileWordList) total search time: 1ms
    WordListDictionary (FileWordList) avg time per search: 105ns
    WordListDictionary (MemoryMappedFileWordList) time to construct: 102ms
    WordListDictionary (MemoryMappedFileWordList) total search time: 0ms
    WordListDictionary (MemoryMappedFileWordList) avg time per search: 63ns

    Test Run #8
    WordListDictionary (FileWordList) time to construct: 94ms
    WordListDictionary (MemoryMappedFileWordList) time to construct: 97ms
    WordListDictionary (FileWordList) total search time: 1ms
    WordListDictionary (FileWordList) avg time per search: 108ns
    WordListDictionary (MemoryMappedFileWordList) total search time: 0ms
    WordListDictionary (MemoryMappedFileWordList) avg time per search: 67ns

    Test Run #9
    WordListDictionary (FileWordList) time to construct: 80ms
    WordListDictionary (FileWordList) total search time: 1ms
    WordListDictionary (FileWordList) avg time per search: 126ns
    WordListDictionary (MemoryMappedFileWordList) time to construct: 85ms
    WordListDictionary (MemoryMappedFileWordList) total search time: 0ms
    WordListDictionary (MemoryMappedFileWordList) avg time per search: 69ns

    Test Run #10
    WordListDictionary (FileWordList) time to construct: 80ms
    WordListDictionary (FileWordList) total search time: 1ms
    WordListDictionary (FileWordList) avg time per search: 113ns
    WordListDictionary (MemoryMappedFileWordList) time to construct: 89ms
    WordListDictionary (MemoryMappedFileWordList) total search time: 0ms
    WordListDictionary (MemoryMappedFileWordList) avg time per search: 59ns

### CODE REVIEW

File: src/main/java/org/passay/dictionary/AbstractFileWordList.java [hunk 1]
```diff
@@ -299,4 +279,108 @@ private boolean hasRemaining()
...
+
+      if (cacheSize > Integer.MAX_VALUE) {
+        throw new IllegalArgumentException("Cache limit exceeded. Try reducing cacheSize.");
+      }
+      map = ByteBuffer.allocateDirect((int) cacheSize).asLongBuffer();
+    }
+
+
+    /**
+     * Puts an entry that maps the word at given index to the byte offset in into the backing file.
```
Comment (dfish3r) [inline src/main/java/org/passay/dictionary/AbstractFileWordList.java]: This should be updated to document the fact that put only succeeds based on the modulus.

File: src/main/java/org/passay/dictionary/AbstractFileWordList.java [hunk 2]
```diff
@@ -354,33 +354,52 @@ void put(final int index, final long position)
...
+     * index.
      *
      * @param  index  Word at index.
      *
      * @return  Nearest cache entry for given index.
      */
     Entry get(final int index)
     {
-      if (modulus == 0) {
+      if (modulus == 0 || index < modulus) {
```
Comment (serac) [inline src/main/java/org/passay/dictionary/AbstractFileWordList.java]: I assume this is the fix to the couple tests that were failing. I don't see how that would matter, but the proof of the pudding is in the eating I suppose.

**Comment by serac** [comment]
Thanks for that commit to fix failing tests. I do like the dedicated `resize` method.

File: src/main/java/org/passay/dictionary/AbstractFileWordList.java (hunk 2 above)
Comment (dfish3r) [inline src/main/java/org/passay/dictionary/AbstractFileWordList.java]: Nope.

Comment (dfish3r) [inline src/main/java/org/passay/dictionary/AbstractFileWordList.java]: See line 138.

Comment (dfish3r) [inline src/main/java/org/passay/dictionary/AbstractFileWordList.java]: This was just a minor optimization.

**Comment by serac** [comment]
Last commit looks good. I believe this is ready to go.

**Pull request merged**